third_party/move/move-model/src/builder/exp_builder.rs

```diff
@@ -2813,112 +2825,281 @@ impl<'env, 'translator, 'module_translator> ExpTranslator<'env, 'translator, 'mo
+            .collect_vec();
+        let mut in_order_fields = BTreeSet::new();
+        if !permutation.is_empty() {
+            let mut prev = permutation[0];
+            in_order_fields.insert(0);
+            for (i, cur) in permutation.into_iter().enumerate().skip(1) {
+                if cur < prev {
+                    continue;
+                } else if cur == prev + 1 {
+                    in_order_fields.insert(i);
+                    prev = cur;
+                } else {
+                    // cur > prev + 1
+                    prev = cur;
+                    in_order_fields = BTreeSet::new();
+                    in_order_fields.insert(i);
+                }
+            }
+        }
+        in_order_fields
+    }
+
+    // given pack<S>{ f_p(1): e_1, ... }, where p is a permutation of the fields
+    // return the struct id of S,
+    // (x, e) ...
+    // arg ...
+    // such that
+    // { let x = e; ...; pack(arg, ...) }
+    // can be the result of the transformation
+    fn translate_exp_fields(
+        &mut self,
+        loc: &Loc,
+        maccess: &EA::ModuleAccess,
+        generics: &Option<Vec<EA::Type>>,
+        fields: &EA::Fields<EA::Exp>,
+    ) -> Option<(
+        QualifiedInstId<StructId>,
+        Vec<(Pattern, ExpData)>,
+        Vec<ExpData>,
+    )> {
+        let struct_name = self.parent.module_access_to_qualified(maccess);
+        let struct_name_loc = self.to_loc(&maccess.loc);
+        let generics = generics.as_ref().map(|ts| self.translate_types(ts));
+        let struct_entry = self.get_struct_report_undeclared(&struct_name, &struct_name_loc)?;
+        let instantiation =
+            self.make_instantiation_or_report(loc, struct_entry.type_params.len(), generics)?;
+        let field_decls = self
+            .get_field_decls_for_pack_unpack(&struct_entry, &struct_name, &struct_name_loc)?
+            .clone();
+
+        self.check_missing_or_undeclared_fields(loc, struct_name, &field_decls, fields)?;
+
+        let in_order_fields = self.in_order_fields(&field_decls, fields);
+        // maps i to (x_i, e_i) where i is the expression idx, i.e., idx of the field exp in the pack exp
+        let mut bindings = BTreeMap::new();
+        // maps i to x_i where i is the field definition idx, i.e., idx of the field in the struct definition
+        let mut args = BTreeMap::new();
+        for (_, name, (exp_idx, field_exp)) in fields.iter() {
+            let field_name = self.symbol_pool().make(name);
+            let (_, def_idx, field_ty) = field_decls.get(&field_name).unwrap();
+            let field_ty = field_ty.instantiate(&instantiation);
+            let translated_field_exp = self.translate_exp(field_exp, &field_ty);
+            if in_order_fields.contains(def_idx) {
+                args.insert(*def_idx, translated_field_exp);
+            } else {
+                // starts with $ for internal generated vars
+                let var_name = self
```

Comment:
Please use some helper functions to build these AST elements.  Any set of k let statements with just 1 output var and <k input vars would be better to put in a separate function.  Maybe even k lines of let statements.

Reply:
Actually this is Rust, so you can return a tuple.

Btw, better if your variable names can be related to in the comment above.  "var" = "x" and "arg" = "exp"?

Reply:
> Btw, better if your variable names can be related to in the comment above. "var" = "x" and "arg" = "exp"?
I added some comments for the vars.

Reply:
> Please use some helper functions to build these AST elements. Any set of k let statements with just 1 output var and <k input vars would be better to put in a separate function. Maybe even k lines of let statements.
I refactored the if branch; but for the else branch, I actually prefer the current.